### PR TITLE
VMU driver support for date/time and buttons + documentation update

### PR DESCRIPTION
**Pull request by gyrovorbis**
I have dedicated a substantial portion of my life trying to understand the VMU and everything it has to offer in an attempt to provide perfect emulation and a reliable developer debugging environment for ElysianVMU. Through reverse engineering the BIOS and working with VMU developers, we have discovered quite a bit!

I couldn't help but notice the fact that we have some missing pieces in the KOS Maple API... I really wanted to make sure KOS is capable of doing everything the VMU has to offer, and hopefully flesh out the API with some really good documentation and corresponding examples. :)

1. Added a higher-level convenience wrapper, vmu_beep(), around vmu_beep_raw(), for generating buzzer square waves
2. Added support for setting and getting the datetime structure from the VMU
3. Added support for polling VMU button states
    - VMU buttons states are actually maintained and can be reported via the Maple protocol when the VMU is plugged in. There now exist 3rd party controllers, such as the RetroFighters StrikerDC gamepad, which expose these buttons and can be used for homebrew games
    - to avoid unnecessary Maple overhead for games that don't care, the buttons have to be enabled via vmu_set_buttons_enabled(), which dynamically adds the periodic polling callback to the driver
    - to further reduce unnecessary overhead, only the first unit in each controller is polled, since the second is assumed to be inaccessible
4. Updated Doxygen 
5. Added support for querying VMU color and icon from the root block

### CODE REVIEW

**Comment by gyrovorbis** [comment]
SHE IS DONE!!!!

I tested the crap out of everything with various controllers, made sure I didn't break anything, made sure the new stuff works, and went through all of the Doxygen that I redid confirming it all looks good:
Screenshot from 2023-08-19 08-47-41
Screenshot from 2023-08-19 08-48-40

File: kernel/arch/dreamcast/hardware/maple/vmu.c [hunk 1]
```diff
@@ -17,19 +25,122 @@
...
+    bt->tm_hour = dt->hour;
+    bt->tm_mday = dt->day;
+    bt->tm_mon  = dt->month - 1;
+    bt->tm_year = dt->year - 1900;
+    bt->tm_wday = dt->weekday != 6? dt->weekday + 1 : 0;
+}
 
-   Thanks to Marcus Comstedt for VMU/Maple information.
- */
+static void vmu_datetime_from_tm(vmu_datetime_t* dt, const struct tm* bt) {
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/maple/vmu.c:50]:
```suggestion
static void vmu_datetime_from_tm(vmu_datetime_t *dt, const struct tm *bt) {
```

File: kernel/arch/dreamcast/hardware/maple/vmu.c [hunk 2]
```diff
@@ -17,19 +25,122 @@
...
+    dt->weekday = bt->tm_wday? dt->weekday - 1 : 6;
+}
 
 static int vmu_attach(maple_driver_t *drv, maple_device_t *dev) {
     (void)drv;
     dev->status_valid = 1;
     return 0;
 }
 
+static void vmu_poll_reply(maple_frame_t* frm) {
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/maple/vmu.c:66]:
```suggestion
static void vmu_poll_reply(maple_frame_t *frm) { 
```

File: kernel/arch/dreamcast/hardware/maple/vmu.c [hunk 3]
```diff
@@ -17,19 +25,122 @@
...
+        /* Verify the size of the frame and grab a pointer to it */
+        //assert(resp->data_len == 7);
+        raw = (vmu_cond_t *)(respbuf + 1);
+
+        /* Fill the "nice" struct from the raw data */
+        cooked = (vmu_state_t *)(frm->dev->status);
+        /* Invert raw struct as nice struct */
+        cooked->buttons = ~(*raw);
+
+        const maple_device_t *cont = maple_enum_dev(frm->dev->port, 0);
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/maple/vmu.c:97]:
```suggestion
        const maple_device_t *cont = maple_enum_dev(frm->dev->port, 0);

        /* Verify the size of the frame and grab a pointer to it */
        //assert(resp->data_len == 7);
        raw = (vmu_cond_t *)(respbuf + 1);

        /* Fill the "nice" struct from the raw data */
        cooked = (vmu_state_t *)(frm->dev->status);
        /* Invert raw struct as nice struct */
        cooked->buttons = ~(*raw);
```

File: kernel/arch/dreamcast/hardware/maple/vmu.c [hunk 4]
```diff
@@ -17,19 +25,122 @@
...
+        frm->dev->status_valid = 1;
+    }
+}
+
+static int vmu_poll(maple_device_t *dev) {
+    uint32_t *send_buf;
+
+    /* Only query for button input on the front VMU of each controller. */
+    if(dev->unit == 1) {
+
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/maple/vmu.c:120]:
```suggestion
```

File: kernel/arch/dreamcast/hardware/maple/vmu.c [hunk 5]
```diff
@@ -181,14 +306,14 @@ int vmu_draw_lcd(maple_device_t * dev, const void *bitmap) {
 
     /* Reset the frame */
     maple_frame_init(&dev->frame);
-    send_buf = (uint32 *)dev->frame.recv_buf;
+    send_buf = (uint32_t*)dev->frame.recv_buf;
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/maple/vmu.c:309]:
```suggestion
    send_buf = (uint32_t *)dev->frame.recv_buf;
```

File: kernel/arch/dreamcast/hardware/maple/vmu.c [hunk 6]
```diff
@@ -327,11 +467,12 @@ static void vmu_block_write_callback(maple_frame_t *frm) {
     /* Wakey, wakey! */
     genwait_wake_all(frm);
 }
-static int vmu_block_write_internal(maple_device_t * dev, uint16 blocknum, uint8 *buffer) {
-    maple_response_t    *resp;
-    int         rv, phase, r;
-    uint32          *send_buf;
-    uint32          blkid;
+
+static int vmu_block_write_internal(maple_device_t * dev, uint16_t blocknum, const uint8_t *buffer) {
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/maple/vmu.c:471]:
```suggestion
static int vmu_block_write_internal(maple_device_t *dev, uint16_t blocknum, const uint8_t *buffer) {
```

File: kernel/arch/dreamcast/hardware/maple/vmu.c [hunk 7]
```diff
@@ -433,7 +574,7 @@ static int vmu_block_write_internal(maple_device_t * dev, uint16 blocknum, uint8
 
 /* Sometimes a flaky or stubborn card can be recovered by trying a couple
    of times... */
-int vmu_block_write(maple_device_t * dev, uint16 blocknum, uint8 *buffer) {
+int vmu_block_write(maple_device_t *dev, const uint16_t blocknum, const uint8_t *buffer) {
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/maple/vmu.c:577]: Why is this `const`?
```suggestion
int vmu_block_write(maple_device_t *dev, uint16_t blocknum, const uint8_t *buffer) {
```

File: kernel/arch/dreamcast/hardware/maple/vmu.c [hunk 8]
```diff
@@ -444,24 +585,124 @@ int vmu_block_write(maple_device_t * dev, uint16 blocknum, uint8 *buffer) {
...
 }
 
-/* Utility function which sets the icon on all available VMUs
-   from an Xwindows XBM. Imported from libdcutils. */
-void vmu_set_icon(const char *vmu_icon) {
-    int i = 0;
-    maple_device_t * dev;
-    uint8   bitmap[48 * 32 / 8];
+int vmu_set_datetime(maple_device_t *dev, time_t unix) {
+    uint32_t *send_buf;
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/maple/vmu.c:597]:
```suggestion
    uint32_t *send_buf;
    struct tm *btime;
```

File: kernel/arch/dreamcast/hardware/maple/vmu.c [hunk 9]
```diff
@@ -444,24 +585,124 @@ int vmu_block_write(maple_device_t * dev, uint16 blocknum, uint8 *buffer) {
...
 
-    vmu_xbm_to_bitmap(bitmap, vmu_icon);
+    assert(dev);
 
-    while((dev = maple_enum_type(i++, MAPLE_FUNC_LCD))) {
-        vmu_draw_lcd(dev, bitmap);
+    struct tm* btime = localtime(&unix);
+    assert(btime); /* If you're failing here, that was an invalid unix timestamp */
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/maple/vmu.c:602]:
```suggestion
    btime = localtime(&unix);
    assert(btime); /* A failure here means an invalid unix timestamp was given. */
```

File: kernel/arch/dreamcast/hardware/maple/vmu.c [hunk 10]
```diff
@@ -444,24 +585,124 @@ int vmu_block_write(maple_device_t * dev, uint16 blocknum, uint8 *buffer) {
...
+    /* Lock the frame */
+    if(maple_frame_lock(&dev->frame) < 0)
+        return MAPLE_EAGAIN;
+
+    /* Reset the frame */
+    maple_frame_init(&dev->frame);
+    send_buf = (uint32_t *)dev->frame.recv_buf;
+    send_buf[0] = MAPLE_FUNC_CLOCK;
+    send_buf[1] = 0;
+    vmu_datetime_from_tm((vmu_datetime_t*)(send_buf + 2), btime);
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/maple/vmu.c:613]:
```suggestion
    vmu_datetime_from_tm((vmu_datetime_t *)(send_buf + 2), btime);
```

File: kernel/arch/dreamcast/include/dc/maple/vmu.h [hunk 11]
```diff
@@ -50,129 +115,478 @@ int vmu_use_custom_color(maple_device_t * dev, int enable);
...
+                                4. duty cycle of square wave 2 (ignored by
+                                   standard mono VMUs) 
+
+    \retval MAPLE_EOK       On success.
+    \retval MAPLE_EAGAIN    If the command couldn't be sent. Try again later.
+    \retval MAPLE_ETIMEOUT  If the command timed out while blocking.
+
+    \sa vmu_beep_waveform
+*/
+int vmu_beep_raw(maple_device_t* dev, uint32_t beep);
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/maple/vmu.h:388]:
```suggestion
int vmu_beep_raw(maple_device_t *dev, uint32_t beep);
```

**Review by ljsebald — CHANGES_REQUESTED** [review]
Got a few suggestions. Just tidying things up a little bit.

File: kernel/arch/dreamcast/hardware/maple/vmu.c (hunk 7 above)
Comment (gyrovorbis) [reply kernel/arch/dreamcast/hardware/maple/vmu.c:577]: Good call, there's no point in making a value arg like that const. Not sure wtf I did.

**Review by ljsebald — APPROVED** [review]

**Pull request merged**